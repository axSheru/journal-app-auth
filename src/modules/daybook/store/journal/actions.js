import journalApi from '@/api/journalApi';

/* export const myAction = async ({ commit }) => {
    
} */

export const loadEntries = async ({ commit }) => {

    const { data } = await journalApi.get('/entries.json')

    if( !data ) {
        commit( 'setEntries', [] )
        return
    }

    const entries = []
    for( let id of Object.keys( data ) ) {
        entries.push({
            id,
            ...data[id]
        })
    }

    commit( 'setEntries', entries )

}

export const updateEntry = async ({ commit }, entry) => {
    
    const { text, date, picture } = entry
    let dataToSave = {text, date, picture}

    await journalApi.put(`/entries/${entry.id}.json`, dataToSave)

    dataToSave.id = entry.id

    commit( 'updateEntry', { ...dataToSave } )

}

export const createEntry = async ({ commit }, entry) => {

    const { text, date, picture } = entry
    let dataToSave = { text, date, picture }

    const { data } = await journalApi.post(`/entries.json`, dataToSave)

    const { name } = data
    dataToSave.id = name

    commit( 'addEntry', { ...dataToSave } )

    return name
    
}

export const deleteEntry = async ({ commit }, id) => {

    await journalApi.delete(`/entries/${ id }.json`)
    commit( 'deleteEntry', id )

}