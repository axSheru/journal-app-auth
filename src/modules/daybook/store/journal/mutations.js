/* export const myMutation = ( state ) => {
    
} */

export const setEntries = ( state, entries ) => {
    
    state.entries = [ ...state.entries, ...entries ]
    state.isLoading = false

}

export const updateEntry = ( state, entry ) => {

    const idx = state.entries.map( e => e.id ).indexOf( entry.id )
    state.entries[idx] = entry
    
}

export const addEntry = ( state, entry ) => {

    state.entries = [ entry, ...state.entries ]//Soluci贸n profe. Se recomienda utilizar el operador spread (...) para romper relaci贸n con el objeto original.
    //state.entries.unshift( entry )//Mi soluci贸n.

    //return entry.id
}

export const deleteEntry = ( state, id ) => {
    
    /* const idx = state.entries.map( e => e.id ).indexOf( id )
    state.entries.splice( idx, 1 ) *///Mi soluci贸n.

    state.entries = state.entries.filter( entry => entry.id !== id )

}