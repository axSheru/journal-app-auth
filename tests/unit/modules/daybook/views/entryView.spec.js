import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import Swal from 'sweetalert2'

import journal from '@/modules/daybook/store/journal'
import EntryView from '@/modules/daybook/views/EntryView'

import { journalState } from '../../../mock-data/test-journal-state'

const createVuexStore = ( initialState ) =>
    createStore({
        modules: {
            journal: {
                ...journal,
                state: { ...initialState }//Se hace de esta forma para poder establecer un estado inicial personalizado.
            }
        }
    })

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoadig: jest.fn(),
    close: jest.fn()
}))

describe('Pruebas en el EntryView', () => {

    const store = createVuexStore( journalState )
    store.dispatch = jest.fn()
    const mockRouter = {
        push: jest.fn()
    }

    let wrapper 

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: '-MjrZSR1hwEC8mfMROiA'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })
    
    test('Debe de sacar al usuario porque el id no existe.', () => {
        
        const wrapper = shallowMount( EntryView, {
            props: {
                id: 'Este ID no existe.'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })

        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })

    })
    
    test('Debe de mostrar la entrada correctamente.', () => {
        
        expect( wrapper.html() ).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()

    })
    
    test('Debe de borrar la entrada y salir.', (done) => {

        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }) )
        
        wrapper.find('.btn-danger').trigger('click')

        expect( Swal.fire ).toHaveBeenCalledWith({
            title: '¿Está seguro?',
            text: 'Una vez borrado no se puede recuperar.',
            showDenyButton: true,
            confirmButtonText: 'Sí, estoy seguro.'
        })

        setTimeout(() => {
            
            expect( store.dispatch ).toHaveBeenCalledWith('journal/deleteEntry', '-MjrZSR1hwEC8mfMROiA')
            expect( mockRouter.push ).toHaveBeenCalled()
            done()

        }, 1)

    })
    
})
