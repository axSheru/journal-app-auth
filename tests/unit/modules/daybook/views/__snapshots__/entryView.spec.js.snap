// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView Debe de mostrar la entrada correctamente. 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">18</span><span class="mx-1 fs-3">Septiembre</span><span class="mx-2 fs-4 fw-light">2021, Sábado</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedió hoy?">
            </textarea></div><img src="https://res.cloudinary.com/axsheru/image/upload/v1632117386/bchmkui2wgfjnqob7ahb.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
